pageserver/src/tenant/timeline/detach_ancestor.rs

```diff
@@ -116,6 +130,472 @@ impl Default for Options {
+        use ExistingAttempt::*;
+        match self {
+            ReadFromIndexPart => false,
+            ContinuedOverRestart(x) | Actual(x, _) if x == *other => true,
+            _ => false,
+        }
+    }
+}
+
+impl ExistingAttempt {
+    fn start_new(&mut self, detached: &TimelineId) -> Result<completion::Completion, Error> {
+        use ExistingAttempt::*;
+        match self {
+            ReadFromIndexPart => {}
+            Actual(other, barrier) if barrier.is_ready() => {
+                if other != detached {
+                    tracing::warn!(prev=%other, next=%detached, "switching ongoing detach; this is not expected to happen normally, but doesn't necessarily mean anything catastrophic");
+                }
+            }
+            Actual(other, _) | ContinuedOverRestart(other) => {
+                return Err(Error::OtherTimelineDetachOngoing(*other))
+            }
+        }
+
+        let (guard, attempt) = Self::new(detached);
+        *self = attempt;
+        Ok(guard)
+    }
+
+    fn new(detached: &TimelineId) -> (completion::Completion, Self) {
+        let (guard, barrier) = completion::channel();
+        (guard, ExistingAttempt::Actual(*detached, barrier))
+    }
+}
+
+/// Represents an across tenant reset exclusive single attempt to detach ancestor.
+#[derive(Debug)]
+pub(crate) struct Attempt {
+    timeline_id: TimelineId,
+
+    _guard: completion::Completion,
+    gate_entered: Option<utils::sync::gate::GateGuard>,
+}
+
+impl Attempt {
+    pub(crate) fn before_shutdown(&mut self) {
+        let taken = self.gate_entered.take();
+        assert!(taken.is_some());
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct SharedStateBuilder {
+    inprogress: std::collections::HashSet<TimelineId>,
+}
+
+impl SharedStateBuilder {
+    /// While loading, visit a timelines persistent [`crate::tenant::IndexPart`] and record if it is being
+    /// detached.
+    pub(crate) fn record_loading_timeline(
+        &mut self,
+        timeline_id: &TimelineId,
+        index_part: &crate::tenant::IndexPart,
+    ) {
+        if index_part
+            .gc_blocking
+            .as_ref()
+            .is_some_and(|b| b.blocked_by_detach_ancestor())
+        {
+            // if the loading a timeline fails, tenant loading must fail as it does right now, or
+            // something more elaborate needs to be done with this tracking
+            self.inprogress.insert(*timeline_id);
+        }
+    }
+
+    /// Merge the loaded not yet deleting in-progress to the existing datastructure.
+    pub(crate) fn build(self, target: &SharedState) {
+        let mut g = target.inner.lock().unwrap();
+
+        assert_eq!(g.latest.is_none(), g.known_ongoing.is_empty());
+
+        g.known_ongoing.extend(self.inprogress);
+        if g.latest.is_none() && !g.known_ongoing.is_empty() {
+            g.latest = Some((ExistingAttempt::ReadFromIndexPart, false));
+            target.gc_waiting.notify_one();
```

Comment:
Why wake the sleeping gc here?  => comment

Reply:
We are waking it up to re-witness the `latest`, after writing over the witnessed (second element in the tuple).

Reply:
witness recording no longer exists because I used async mutex per Johns idea in #8600, which essentially replaced it for all of the notifies (I failed to see it).